File: Snake_ladder.cpp
#include<bits/stdc++.h>
#include<time.h>
using namespace std;
int main()
{
    srand(time(0));
    int snk, lad, per; // Number of snakes, ladders and persons
    int ip, fp; // Initial and final position
    cin >> snk; // taking input of number of snakes
    unordered_map<int, int> snakes; //int to int map to find snake address
    for(int i=0;i<snk;i++)
    {
        cin >> ip >> fp;
        snakes[ip] = fp;
    }
    cin >> lad; // taking input of number of ladders
    unordered_map<int, int> ladders; //int to int map to find ladder address
    for(int i=0;i<lad;i++)
    {
        int ip, fp;
        cin >> ip >> fp;
        ladders[ip] = fp;
    }
    cin >> per;
    vector<pair<string, int>> person; //string to int vector to keep count of present address of persons
    for(int i=0;i<per;i++)
    {
        string name;
        cin >> name;
        person.push_back(make_pair(name,0));
    }
    int c = 0;
    while(1)
    {
        int x = (rand()%6)+1; // Random dice value generator (1-6)
        int initial;
        int final = person[c].second + x;
        /*Checking whether at next position a snake exist or not*/
        if(final<=100 && snakes.find(final)!=snakes.end())
        {
            initial = person[c].second;
            person[c].second = snakes[final];
            cout << person[c].first << " rolled a " << x << " and moved from " << initial << " to " << person[c].second << endl;
        }
        /*Checking whether at next position a ladder exist or not*/
        else if(final<=100 && ladders.find(final)!=ladders.end())
        {
            initial = person[c].second;
            person[c].second = ladders[final];
            cout << person[c].first << " rolled a " << x << " and moved from " << initial << " to " << person[c].second << endl;
            if(person[c].second==100)
            {
                cout << person[c].first << " wins the game" << endl;
                break;
            }
        }
        /* if neither snake nor ladder exist then moving x moves forward*/
        else
        {
            if(final<=100)
            {
                initial = person[c].second;
                person[c].second+=x;
            }
            else
            {
                initial = person[c].second;
                
            }
            cout << person[c].first << " rolled a " << x << " and moved from " << initial << " to " << person[c].second << endl;
            if(person[c].second==100)
            {
                cout << person[c].first << " wins the game" << endl;
                break;
            }
        }
        /* Printing the initial and final position for each player*/
        c = (c+1)%per;
    }
    return 0;
}

File: snake-ladder.py
import random
def main():
    snakes=int(input())
    head=[]
    tail=[]
    for i in range(snakes):
        s_head,s_tail=map(int,input().split())
        head.append(s_head)
        tail.append(s_tail)
    ladders=int(input())
    start=[]
    end=[]
    for i in range(ladders):
        l_start,l_end=map(int,input().split())
        start.append(l_start)
        end.append(l_end)
    names=[]
    players=dict()
    p=int(input())
    for k in range(p):
        name=input()
        names.append(name)
        players[name]=0
    x=0
    while True:
        dice=random.randrange(1,7)
        curr_pos=players[names[x]]+dice
        if curr_pos in head:
            curr_index=head.index(curr_pos)
            curr_pos=tail[curr_index]
        elif curr_pos>100:
            continue
        elif curr_pos in start:
            curr_index=start.index(curr_pos)
            curr_pos=end[curr_index]
        elif curr_pos==100:
            print(names[x],"rolled a",dice,"and moved from",players[names[x]],"to",curr_pos)
            print(names[x],"wins the game")
            break
        print(names[x],"rolled a",dice,"and moved from",players[names[x]],"to",curr_pos)
        players[names[x]]=curr_pos
        x+=1
        if x==p:
            x=0
            
if __name__== "__main__":
  main()